### ISSUE TITLE
Eager-loading for Address data

### ISSUE THREAD

**Issue by lindseydiloreto** [doublesecretagency/craft-googlemaps#12]
This one is a bit tricky. The native behavior for [eager loading](https://craftcms.com/docs/3.x/dev/eager-loading-elements.html) only takes into account full-fledged **Elements**. However, the Address is just a single field.

Assuming that it's even possible, we'll probably need to get a bit creative to make eager-loading work. We're not ready to take this on quite yet, but in theory we'll circle back on it eventually.

We'll keep it on our radar, but definitely can't provide an ETA. Feel free to subscribe to this thread for eventual updates.

---

> Migrated from Smart Map: https://github.com/doublesecretagency/craft-smartmap/issues/15

**Comment by bencroker** [comment]
I came across this while encountering an _N+1_ problem when listing elements along with their addresses and wondered if there was an eager-loading opportunity. I'd suggest approaching this with a _behavior_ and a _left join_ on the addresses table, similar to how I achieved it here:
https://github.com/putyourlightson/craft-entry-count/blob/82f92377178de7b1c4cf7c6ef36997f5d10abb72/src/EntryCount.php#L97-L132

**Comment by lindseydiloreto** [comment]
Ah, fantastic tip! Thanks @bencroker, I'll almost certainly go this route when I eventually implement it. 🍺

It may still be awhile, however, before I make any adjustments. Craft 4 is likely to bring some changes which may affect this architecture.

**Comment by MoritzLost** [comment]
I just ran across the _N+1_ problem with the address field as well. I'm working on an API that returns a couple hundred entries with an address field. Every entry needs one additional query for the address, I'm trying to optimize that. I actually found a mostly workable solution that would require only a minor change without a backwards compatibility break.

First, here's the current code that's causing the N+1 problem (`location` being the address field):

```php
$coordinates = ['lat' => 50.9549, 'lng' => 7.0121, ];

$entries = Entry::find()
    ->location(['target' => $coordinates])
    ->orderBy(['distance' => SORT_ASC, 'title' => SORT_ASC])
    ->limit(100)
    ->all();

foreach ($entries as $entry) {
    $location = $entry->location;
    // do additional stuff
}
```

Now every call to `$entry->location` causes one additional database query because the location record is loaded on demand in `AddressField::normalizeValue`:

```php
$record = AddressRecord::findOne([
    'elementId' => $element->id,
    'fieldId' => $this->id,
]);
```

To prevent this, we can just use the address active record class directly to query all addresses at once:

```php
$locationsField = Craft::$app->getFields()->getFieldByHandle('location');
$locations = Address::find()->where([
    'elementId' => array_column($entries, 'id'),
    'fieldId' => $locationsField->id,
])->indexBy('elementId')->all();

foreach ($entries as $entry) {
    $location = $locations[$entry->id];
    // do additional stuff
}
```

The problem is that now, `$location` is an `doublesecretagency\googlemaps\records\Address` instead of an `doublesecretagency\googlemaps\models\Address`. Now I can just copy over everything that `AddressField::normalizeValue` does to create a model based on the record:

```php
// Get the record attributes
$omitColumns = ['dateCreated','dateUpdated','uid'];
$attr = $record->getAttributes(null, $omitColumns);

// Convert coordinates to floats
$attr['lat'] = ($attr['lat'] ? (float) $attr['lat'] : null);
$attr['lng'] = ($attr['lng'] ? (float) $attr['lng'] : null);

// Check if JSON is valid
// Must use this function to validate (I know it's redundant)
$valid = json_decode($attr['raw']);

// Convert raw data to an array
$attr['raw'] = ($valid ? Json::decode($attr['raw']) : null);

// If part of a proximity search, get the distance
if ($value && is_numeric($value)) {
    $attr['distance'] = (float) $value;
}

// Return an Address model
return new AddressModel($attr);
```

This works fine, but of course copying all this code over is not ideal. I see two possible solutions:

- In `AddressField::normalizeValue`, check if the passed `$value` is already an address record and then perform the same steps to transform it into an address model. Then we could just call `AddressField::normalizeValue($addressRecord)` to get the address model.
- Move the logic to map an address record to an address model into a public static method on the `AddressField` or a helper class. Then I could call, for example, `AddressField::addressRecordToModel($locationRecord)`.

Maybe even both. Both of those would be backwards compatible. Maybe the process could even be simplified by providing a utility to perform the bulk loading of addresses.

@lindseydiloreto Would you be open to those suggestions? I'd be happy to write up a PR.

**Comment by DieterVanhove** [comment]
Hi all 👋

Just checking in to see if there’s been any progress on this feature or if it’s currently on the roadmap. I’m running into the same N+1 issue on an overview page listing studios, where each studio has an address field. As expected, the address data isn’t being eager-loaded, which results in a separate query for each entry.

I’ve reviewed the suggestions above and understand the current workarounds, but I was wondering if any development has been done in this area or if there are plans to address it more formally in an upcoming release?

Thanks in advance, and appreciate all the hard work on this plugin!

**Comment by lindseydiloreto** [comment]
Hi @DieterVanhove, thanks for checking in. This issue is technically still on our radar, though not currently at the top of the list.

I most likely missed the window to take @MoritzLost up on his offer of a PR (sorry Moritz, thanks anyway).

We'll keep this thread open until we eventually tackle it!

**Comment by MoritzLost** [comment]
@lindseydiloreto Yeah, at the moment I don't have a minute to spare, unfortunately.

I think an 80/20 solution would be relatively straightforward, though. Moving the logic for creating the `Address` model to a publicly available helper method would enable my solution above to work without too much boilerplate.
